sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -19,38 +19,75 @@ package org.apache.spark.sql.execution.datasources.v2
 import org.apache.spark.sql.connector.catalog.{StagingTableCatalog, TableCapability}
+import org.apache.spark.sql.connector.read.V1Scan
 import org.apache.spark.sql.connector.read.streaming.{ContinuousStream, MicroBatchStream}
-import org.apache.spark.sql.execution.{FilterExec, ProjectExec, SparkPlan}
+import org.apache.spark.sql.execution.{FilterExec, LeafExecNode, ProjectExec, RowDataSourceScanExec, SparkPlan}
 import org.apache.spark.sql.execution.datasources.DataSourceStrategy
 import org.apache.spark.sql.execution.streaming.continuous.{ContinuousCoalesceExec, WriteToContinuousDataSource, WriteToContinuousDataSourceExec}
+import org.apache.spark.sql.sources.TableScan
 import org.apache.spark.sql.util.CaseInsensitiveStringMap
 
-object DataSourceV2Strategy extends Strategy with PredicateHelper {
+class DataSourceV2Strategy(session: SparkSession) extends Strategy with PredicateHelper {
 
   import DataSourceV2Implicits._
 
+  private def withProjectAndFilter(
+      project: Seq[NamedExpression],
+      filters: Seq[Expression],
+      scan: LeafExecNode,
+      needsUnsafeConversion: Boolean): SparkPlan = {
+    val filterCondition = filters.reduceLeftOption(And)
+    val withFilter = filterCondition.map(FilterExec(_, scan)).getOrElse(scan)
+
+    if (withFilter.output != project || needsUnsafeConversion) {
+      ProjectExec(project, withFilter)
+    } else {
+      withFilter
+    }
+  }
+
   override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+    case PhysicalOperation(project, filters,
+        relation @ DataSourceV2ScanRelation(_, v1Scan: V1Scan, output)) =>
+      val pushedFilters = relation.getTagValue(V2ScanRelationPushDown.PUSHED_FILTERS_TAG)
+        .getOrElse(Seq.empty)
+      val v1Relation = v1Scan.toV1TableScan(session.sqlContext)
+      if (v1Relation.schema != v1Scan.readSchema()) {
+        throw new IllegalArgumentException(
+          "The fallback v1 relation reports inconsistent schema:\n" +
+            "Schema of v2 scan:     " + v1Scan.readSchema() + "\n" +
+            "Schema of v1 relation: " + v1Relation.schema)
+      }
+      val rdd = v1Relation match {
+        case s: TableScan => s.buildScan()
+        case _ =>
+          throw new IllegalArgumentException(
+            "`V1Scan.toV1Relation` must return a `TableScan` instance.")
+      }
+      val unsafeRowRDD = DataSourceStrategy.toCatalystRDD(v1Relation, output, rdd)
+      val originalOutputNames = relation.table.schema().map(_.name)
+      val requiredColumnsIndex = output.map(_.name).map(originalOutputNames.indexOf)
+      val dsScan = RowDataSourceScanExec(
```

Comment:
It seems like there is quite a bit of work done here in the strategy that should be in the exec node instead, like converting the RDD to a catalyst RDD. What about building a subclass of RowDataSourceScanExec that handles these concerns?

Reply:
This may be a good idea, but this is what we do for ds v1 (see the v1 rule `DataSourceStrategy`). I'd like to avoid changing the existing design choice we made before.

Reply:
I'd like to clean this up before committing this change. I'm not so sure that this was a purposeful design choice in v1. If so, what was the rationale for putting this logic in the strategy?

Reply:
I tried and it's not easy to refactor. The query plans in Spark are all case class and you can't create a sub-class. And it's only a few lines of code that can be encapsulated, the part of getting the v1 scan should belong to the strategy.

If you have some ideas, please help and send a PR to my branch. I'm happy to see approaches that can make the code cleaner.

Reply:
how about an alternate constructor?

Reply:
The Scala auxiliary constructors must call other constructors as the FIRST action, e.g. 
```
def this(...) {
  do something
  this(...)
}
```
is not allowed.

We can create an object and add an `apply` method to encapsulate the 4 lines of code, but does it really matter? The 4 lines of code only appear once here.